### ISSUE TITLE
pilot & deliver team parameters are misnamed

### ISSUE THREAD

**Issue by samritchie** [fastlane/fastlane#5433]
The `pilot` & `deliver` tools have the team-related command line parameters (pilot shown): 

    -q, --team_id [VALUE] The ID of your team if you're in multiple teams (PILOT_TEAM_ID) 
    -r, --team_name STRING The name of your team if you're in multiple teams (PILOT_TEAM_NAME) 

For comparison, `sigh` also has these parameters:

    -b, --team_id STRING The ID of your team if you're in multiple teams (SIGH_TEAM_ID) 
    -l, --team_name STRING The name of your team if you're in multiple teams (SIGH_TEAM_NAME) 

Unfortunately these refer to completely different things - the former refers to the iTunes Connect provider, the latter to the Dev Centre Team. This is enormously confusing and inconsistent. to make matters worse:

1. the error message returned if the wrong value is specified for Team ID is incredibly vague: (`invalid request body`)
2. these parameters are ignored unless the apple ID is connected to more than one iTunes Connect account (which has only been possible recently), meaning an incorrect team ID may work correctly for a long time, and then mysteriously fail when the user is granted access to another account.

As far as I’m aware, Apple don’t use the term ‘Team’ to refer to iTunes Connect - in public documentation they call it an ‘Account’, internally it appears to be labelled a ‘provider’. 

These parameters should be renamed (with a suitable deprecation period) - I suggest something like:

    -q, --itc_account_id [VALUE] The ID of your iTC account if you have multiple accounts (PILOT_ITC_ACCOUNT_ID) 
    -r, --itc_account_name STRING The name of your iTC account if you have multiple accounts (PILOT_ITC_ACCOUNT_NAME) 

and the tools should return a sensible error message if the parameters are incorrect (this may require pre-flighting). It would be great if these were always validated, rather than only if the user has multiple accounts.

Migrated from #1526

**Comment by TKBurner** [comment]
@samritchie Hope you are doing well! Are you interested in submitting a PR for this? I'd love to see that and work with you 👍 Let me know!

**Comment by samritchie** [comment]
@TKBurner I could give it a go. A couple of questions though:

1. How do you deprecate an option (in particular, retain the short switch, but change the key & env var)? If it’s just duplicated, how is the short switch resolved? 
2. It looks like @mfurtak added `PILOT_ITC_PROVIDER` and `DELIVER_ITC_PROVIDER` env-only options; these would appear to duplicate the team_id options, but I’m not really clear on how they’re used?

**Comment by mfurtak** [comment]
Hi @samritchie 👋 

`PILOT_ITC_PROVIDER` and `DELIVER_ITC_PROVIDER` are used to pass a provider name through the `iTMSTransporter`. "provider" is just what the `iTMSTransporter` calls this organizational entity. While this value is commonly equal to the Developer Portal team ID, for teams that have existed a very long time, it is not.

See https://github.com/fastlane/fastlane/pull/4746#issuecomment-223482891 for a way to get the right values for this from Apple.

**Comment by mfurtak** [comment]
@KrauseFx Can you give your thoughts on this ticket?

**Comment by KrauseFx** [comment]
Hey @samritchie,
Thanks for raising this issue, that's super helpful - a very good point! These are the things that are really hard to change, as we have to be backwards compatible to not break any setups with a new release. 

> the error message returned if the wrong value is specified for Team ID is incredibly vague: (invalid request body)
That's far from ideal, and definitely something we should fix, regardless of if we change the parameter names. Is it reproducible on every run? If so, do you feel like you could submit a PR adding a test + nice error message?

> meaning an incorrect team ID may work correctly for a long time, and then mysteriously fail when the user is granted access to another account.
Is that something we can detect? e.g. when a Team ID or Team Name is provided, but the user is only in one iTC team, we should a clearly visible warning? I thought we already do so.

To sum up, I wouldn't change the available parameters, as it would break existing setups, and implementing an alias is a lot of extra work, with not a lot of advantages. Instead let's focus on improving the error messages and warnings, and also the documentation around those parameters (both in the source code, and in the user-facing documentation)

What do you think? 👍

**Comment by samritchie** [comment]
@mfurtak there’s a third ID. Awesome. I’d never seen that provider short name before. 

I’ve read through all the issues & I think I’m managing to grasp what’s going on:

* older versions of pilot/deliver never passed a provider ID through to transporter; immediately post-multi-account iTunes, iTunes Connect used some sort of metadata inspection to locate the appropriate provider.
* pilot/deliver DID require the iTC team ID for multi-account users, however this was still not passed through to transporter (presumably it was just a spaceship thing)
* at a later point, (May/June timeframe?), iTunes Connect started requiring an explicit provider for multi-account users. However, the provider is not specified via its provider ID, because that would be too sensible, but instead uses a short name that’s normally the dev centre ID, but not always.
* the `:dev_portal_team_id` option was added as a workaround, under the assumption that this was always the same as the short name.
* when apparent it wasn’t, the `itc_provider` option was then added to explicitly set the short name for the older accounts.

Well, that’s just spiffy. Ideally fastlane would look up the short name using transporter if it’s not supplied, which should be feasible if the itc provider name or id is specified. I assume `:dev_portal_team_id` can be deprecated?

@KrauseFx:

> That's far from ideal, and definitely something we should fix, regardless of if we change the parameter names. Is it reproducible on every run?

I last tested about a month ago, I assume it’s still in place. I’ll recheck. It looks like a generic message and might require explicit preflight validation to reliably test, I can try and have a go at this, my Ruby is not great though.

> when a Team ID or Team Name is provided, but the user is only in one iTC team, we should a clearly visible warning? I thought we already do so.

It doesn’t look like it (at least, not from pilot). Personally, the logic I’d use is ‘IF the itc team ID or name is supplied, THEN validate against the user iTC teams and print a warning/error if invalid’. Maybe it could be a warning for single-team users and an error for multi-team users; this will cover off the vague error issue as well.

> I wouldn't change the available parameters, as it would break existing setups, and implementing an alias is a lot of extra work, with not a lot of advantages. Instead let's focus on improving the error messages and warnings, and also the documentation around those parameters (both in the source code, and in the user-facing documentation)

I think people tend to underestimate the long-term cost of bad naming - ideally there’d be better support for deprecation/renaming in `ConfigItem` (possibly something like `deprecated_aliases`, `deprecated_short_options`, `deprecated_env_names` etc). However, I don’t have the knowledge or time to do a change this invasive. I can update the parameter descriptions & I’ll have a look at the docs though.

**Comment by KrauseFx** [comment]
> the error message returned if the wrong value is specified for Team ID is incredibly vague: (invalid request body)

This has been solved with the most recent _spaceship_ release 👍
